packages/ui-components/src/features/Form/FormInput.tsx

```diff
@@ -12,17 +12,35 @@ type HookFormInputWrapperProps = Record<string, unknown> & {
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
   Input: ComponentType<any>;
   required?: boolean;
+  nonWhitespace?: boolean;
+  validate?:
+    | ((value: string) => boolean | string)
+    | Record<string, (value: string) => boolean | string>;
 };
 export const FormInput = ({
   name,
   required,
+  nonWhitespace,
+  validate = {},
   options = {},
   Input,
   ...props
 }: HookFormInputWrapperProps) => {
   const { register, errors = {} } = useFormContext();
+
   const requiredConfig = required ? { required: 'Required' } : {};
-  const registerOptions = { ...options, ...requiredConfig };
+
+  const verifyIsNonwhitespace = (value: string) => !!value.trim() || 'Must not be empty';
+  const validateConfig =
+    typeof validate === 'function'
+      ? { validate: nonWhitespace ? { validate, verifyIsNonwhitespace } : { validate } }
+      : { validate: nonWhitespace ? { ...validate, verifyIsNonwhitespace } : { ...validate } };
+
+  const registerOptions = {
+    ...requiredConfig,
+    ...validateConfig,
+    ...options,
+  };
```

Comment:
Would appreciate a second (or third!) opinion on this:

The `required` and `nonWhitespace` props just promote `required` and `validate` “up one level” so they can be passed as props to `FormInput`, rather than inside the `options` object. But this doesn’t stop the developer from providing the same option in both places

If my understanding of TypeScript is correct, the way I’ve implemented it gives `options.required` and `options.validate` higher precedence over the custom `required` and `nonWhitespace` shorthand.

In my head, this is a graceful mode of failure. Do you see any potential issues further down the line?

Reply:
Yeah, no, I don’t see why a dev would intentionally set both. The shorthand is a big neater, but if we need to customise the behaviour or overload something, we still have the option to do so via `options`